[Help center]  Restore unread status

AllTerrainDeveloper opened the pull request:
<!--
Link a related Github/Linear issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword. Consider using "part of" instead.
-->

Part of https://github.com/Automattic/wp-calypso/pull/102667

## Proposed Changes

The[ Pull Request](https://github.com/Automattic/wp-calypso/pull/102667) enabling history for Odie AI Conversations mistakenly removed the unread badges and count due to a lack of knowledge on my side. This set of changes is bringing back that minimal set of changes, with improved namings, and a bit less convoluted development.

## Why are these changes being made?
<!--
It's easy to see what a PR does but much harder to find out why it was made,
particularly when researching old changes in history. Record an explanation of
the motivation behind this change and how it will help.
-->

Happiness Engineers were asking for this feature to be added, and it's indeed a useful indicator for our users. The badge for unread messages is being added as well as the blueish fill color for unread conversations.

## Testing Instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

1. Apply this branch locally, it's a pretty much needed step. Don't use Calypso live link.
2. Modify the files: `help-center-smooch.tsx` and `use-authenticate-zendesk-messaging.ts` (37d87-pb/  37d88-pb/). Those changes will enable the Zendesk production endpoints (not staging)
3. Now, use your ZD account. For that, you will need to give yourself permissions (if you are a developer) through Mission Control -> Tools -> Happiness -> Zendesk Account Manager. Give you permissions for WPCOM, and click on "Just today"
4. Now run Calypso locally and start a Odie AI conversation, escalate the conversation, identify in the conversation as an a11n and kindly ask not to handle this request. Close the Help Center now.
5. Once in Zendesk, handle your own conversation, reply to you a few times.
6. Open now Help Center in WordPress (Calypso local instance)
7. Assert the following:
7.1 The number of unread messages match in the main help center window.
7.2 After clicking on History, the conversation is highlighted in blue
7.3 After clicking on the highlighted conversation, the blueish background is removed.

**Screenshots**
|Step 7.1|Step 7.2 |Step 7.3|
|---|---|---|
|![image](https://github.com/user-attachments/assets/583bae21-cbe2-44ca-a152-3d708a1d3c8c)|![image](https://github.com/user-attachments/assets/dfc6dda8-34c8-473a-9af5-26f189d0b128)|![image](https://github.com/user-attachments/assets/6baac29f-face-4a06-a3c8-eeff06376599)

## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
    - [ ] For UI changes, have we tested the change in various languages (for example, ES, PT, FR, or DE)? The length of text and words vary significantly between languages. 
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-aUh-p2)?

AllTerrainDeveloper committed:
Initial commit: Restore unread status

ehg commented:
I took this for a spin, and couldn't see any of unread functionality. The message in the "Recent conversation" didn't update until I went to the conversation itself.

AllTerrainDeveloper committed:
Enable real time messaging update

AllTerrainDeveloper committed:
Merge branch 'trunk' into Odie/restore-help-center-badge-and-unread-status

escapemanuele commented on packages/help-center/src/components/help-center-smooch.tsx:
```diff
@@ -94,9 +95,10 @@ const HelpCenterSmooch: React.FC< { enableAuth: boolean } > = ( { enableAuth } )
 				return;
```
In my case, when receiving a message, the recent chat is not updated because `lastMessageReceivedAt` is not set, as this function returns

AllTerrainDeveloper replied:
What do you mean? `isHelpCenterShown` or `setLastMessageReceivedAt`?

AllTerrainDeveloper replied:
But this would only work when Help Center is shown, isn't it? I mean, if it's not shown, we don't need to update any badge

AllTerrainDeveloper committed:
Aditional fixes

AllTerrainDeveloper commented:
I have added a set of aditional fixes that might need some explanation.

I have removed some loading guards that were returning null, this was making the screen to flicker, so now it will only return false, while is loading and no previous data. It has sense now, as if we receive a message, we were setting the help center to a blank state on some parts when we still need to show old data to present the new data, instead of the flow "old data -> blank (loading) -> new data"

Another thing was to add timestamps to the query cache, that will force fetching when new data is receied. I deally we should have integrated Smooch differently instead of relying on timestamps, but that would require quite a lot of work, and might be a nice work for a future a11n :)

ehg reviewed (approved):
Code LGTM and it works!

stephanethomas commented on packages/help-center/src/components/help-center-chat-history.tsx:
```diff
@@ -55,7 +55,7 @@ const Conversations = ( {
 		<>
 			{ conversations.map( ( conversation ) => {
-				const lastMessage = getLastMessage( { conversation } );
+				const { lastMessage, unreadMessages } = getLastMessageAndUnread( { conversation } );
```
It feels like the name of that function is missing a bit. Since `getLastMessageAndUnreadMessages` is both long and repetitive, I would suggest a slightly different name:
```suggestion
				const { lastMessage, unreadMessages } = getLastAndUnreadMessages( { conversation } );
```

stephanethomas commented on packages/help-center/src/components/help-center-support-chat-message.tsx:
```diff
@@ -32,10 +32,12 @@ export const HelpCenterSupportChatMessage = ( {
 	message: OdieMessage | ZendeskMessage;
 	sectionName?: string;
 	conversation: OdieConversation | ZendeskConversation;
+	unreadMessages?: number;
```
I would be more explicit here as it's not always clear looking at the code below what that variable contains without checking its type (e.g. one could expect a list of unread messages):
```suggestion
	numberOfUnreadMessages?: number;
```

stephanethomas commented on packages/help-center/src/components/help-center-support-chat-message.tsx:
```diff
@@ -86,6 +88,8 @@ export const HelpCenterSupportChatMessage = ( {
 		);
 	};
 
+	const hasUnread = unreadMessages > 0;
```
Tangentially related to my previous message, and certainly a bit nitpicky but something you may want to consider:
```suggestion
	const hasUnreadMessages = unreadMessages > 0;
```

stephanethomas commented on packages/help-center/src/components/help-center-support-chat-message.tsx:
```diff
@@ -100,11 +104,20 @@ export const HelpCenterSupportChatMessage = ( {
 			} ) }
 		>
-			<div className={ clsx( 'help-center-support-chat__conversation-avatar' ) }>
+			<div
+				className={ clsx( 'help-center-support-chat__conversation-avatar', {
+					'has-unread-messages': unreadMessages > 0,
+				} ) }
+			>
 				{ renderAvatar() }
+
+				{ unreadMessages > 0 && (
+					<div className="help-center-support-chat__conversation-badge">+{ unreadMessages }</div>
+				) }
```
What about using the _boolean_ variable instead of duplicating code 😅?

```suggestion
			<div
				className={ clsx( 'help-center-support-chat__conversation-avatar', {
					'has-unread-messages': hasUnreadMessages,
				} ) }
			>
				{ renderAvatar() }

				{ hasUnreadMessages && (
					<div className="help-center-support-chat__conversation-badge">+{ unreadMessages }</div>
				) }
```

AllTerrainDeveloper replied:
Oops :)

stephanethomas commented on packages/help-center/src/components/utils.tsx:
```diff
@@ -40,14 +43,22 @@ export const generateContactOnClickEvent = (
+			: null;
+
+	const { unreadMessages } = calculateUnread( [ conversation as ZendeskConversation ] );
+	return { lastMessage, unreadMessages };
```
We still call `getLastMessageAndUnread()` to retrieve only the last message half of the time, so I don't see much value in grouping the retrieval of the last message with the computation of the number of messages. What about calling them separately?

If that was me, I would even rename `calculateUnread()` into something more explicit such as `getNumberOfUnreadConversationsAndMessages()`. With the latter it's easy to understand what that function does, with the former I would have to look at the code to see what output I should expect.

stephanethomas commented on packages/odie-client/src/data/use-get-unread-conversations.ts:
```diff
@@ -4,12 +4,18 @@ import { useCallback } from 'react';
+	conversations: Conversation[] | ZendeskConversation[] | undefined | null
+) => {
+	if ( ! Array.isArray( conversations ) || conversations.length === 0 ) {
+		return { unreadConversations: 0, unreadMessages: 0 };
```
A small suggestion related to some previous feedback, if you decide to follow that approach:
```suggestion
		return { numberOfUnreadConversations: 0, numberOfUnreadMessages: 0 };
```

stephanethomas commented on packages/help-center/src/hooks/use-get-history-chats.ts:
```diff
@@ -140,6 +143,7 @@ export const useGetHistoryChats = (): UseGetHistoryChatsResult => {
 		otherSupportInteractions,
 		odieConversations,
 		supportInteractions,
+		lastMessageReceivedAt,
```
We don't use that variable in the `useEffect()` clause, should we still need to define it as a dependency?

AllTerrainDeveloper replied:
We want to reorder the array for new data, but is true that the rest of the deps will change

stephanethomas commented on packages/help-center/src/components/help-center-chat-history.tsx:
```diff
@@ -67,6 +67,7 @@ const Conversations = ( {
 						key={ conversation.id }
 						message={ lastMessage }
 						conversation={ conversation }
+						unreadMessages={ unreadMessages }
```
You can leave it that way if you prefer, but since we already pass the whole `conversation` object, it would be really easy to retrieve the number of unread messages from inside `HelpCenterSupportChatMessage` components - which would simplify any calling code.

AllTerrainDeveloper replied:
From all the suggestions I'm going to leave this one, somehow I like it more modular too

stephanethomas reviewed (commented):
I haven't tested those changes yet but I'd love to submit some feedback first about how we could improve the code in term of readability and duplication. I'll give those changes a try once you've decided if you want to follow those suggestions, otherwise feel free to merge them since Chris has already tested them.

AllTerrainDeveloper committed:
Address Stéphane suggestions

AllTerrainDeveloper committed:
More suggestions

stephanethomas commented on packages/help-center/src/components/help-center-chat-history.tsx:
```diff
@@ -55,6 +59,9 @@ const Conversations = ( {
 		<>
 			{ conversations.map( ( conversation ) => {
+				const { numberOfUnreadMessages } = calculateUnread( [
+					conversation as ZendeskConversation,
```
It's probably me not being familiar enough with TypeScript but what happens when the `conversation` is an `OdieConversation`?

AllTerrainDeveloper replied:
As long as it has enough overlap it's ok. And actually the property is being used in a component that accepts either a `ZendeskConversation` or a `OdieConversation`, so it's ok.

stephanethomas replied:
Well that's exactly my point - do we really need to cast that `conversation` into `ZendeskConversation`? If the conversation is of type `OdieConversation`, will it trigger a runtime error?

AllTerrainDeveloper replied:
Oh well, just checked it. The parameter is actually used in `calculateUnread`. `CalculateUnread` uses a property that `OdieConversation` is missing: `participants`. That's why It's casted here.

`OdieConversation` won't have unread messages as it's always replying immediately and `OdieConversation` not having the participants property, when casted to `ZendeskConversation`, returns 0.

I hope I was able to clarify it enough 😄

stephanethomas commented on packages/help-center/src/components/utils.tsx:
```diff
@@ -40,14 +42,21 @@ export const generateContactOnClickEvent = (
-		? conversation.messages[ conversation.messages.length - 1 ]
-		: null;
+} ): OdieMessage | ZendeskMessage | null => {
+	const lastMessage =
+		Array.isArray( conversation.messages ) && conversation.messages.length > 0
+			? conversation.messages[ conversation.messages.length - 1 ]
+			: null;
+
+	return lastMessage;
```
I believe we could really do without that local variable, it's already pretty clear what that function returns:
```suggestion
	return Array.isArray( conversation.messages ) && conversation.messages.length > 0
		? conversation.messages[ conversation.messages.length - 1 ]
		: null;
```

AllTerrainDeveloper replied:
Right, enough information already in the method name.

stephanethomas reviewed (approved):
I've tested those changes and they work great, time to ship it!

AllTerrainDeveloper committed:
Add Stéphane's suggestion
